packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayer.m

```diff
@@ -82,68 +124,55 @@ - (instancetype)initWithPlayerItem:(AVPlayerItem *)item
   };
   _videoOutput = [avFactory videoOutputWithPixelBufferAttributes:pixBuffAttributes];
 
-  [self addObserversForItem:item player:_player];
+  // Set up all necessary observers to report video events.
+  _playerItemObservations = @{
+    @"loadedTimeRanges" : [NSValue valueWithPointer:timeRangeContext],
+    @"status" : [NSValue valueWithPointer:statusContext],
+    @"presentationSize" : [NSValue valueWithPointer:presentationSizeContext],
+    @"duration" : [NSValue valueWithPointer:durationContext],
+    @"playbackLikelyToKeepUp" : [NSValue valueWithPointer:playbackLikelyToKeepUpContext],
+  };
+  _playerObservations = @{
+    @"rate" : [NSValue valueWithPointer:rateContext],
+  };
+  FVPRegisterKeyValueObservers(self, _playerItemObservations, item);
+  FVPRegisterKeyValueObservers(self, _playerObservations, _player);
+  [[NSNotificationCenter defaultCenter] addObserver:self
+                                           selector:@selector(itemDidPlayToEndTime:)
+                                               name:AVPlayerItemDidPlayToEndTimeNotification
+                                             object:item];
 
   [asset loadValuesAsynchronouslyForKeys:@[ @"tracks" ] completionHandler:assetCompletionHandler];
 
   return self;
 }
 
 - (void)dealloc {
-  if (!_disposed) {
-    [self removeKeyValueObservers];
-  }
+  // In case dispose was never called for some reason, remove any remaining observers to prevent
+  // crashes.
+  FVPRemoveKeyValueObservers(self, _playerItemObservations, _player.currentItem);
+  FVPRemoveKeyValueObservers(self, _playerObservations, _player);
 }
 
 - (void)dispose {
   _disposed = YES;
-  [self removeKeyValueObservers];
 
```

Comment:
not directly related to this PR: can we return early if `_dispose` is already YES, or this method shouldn't be a no-op if called a second time?